### PR TITLE
CI: add automated GitHub prereleases

### LINKED ISSUE #8541: docs: mention GitHub CI build artifacts in README
We often have to reference them, so maybe adding a reference to the README could save us some effort...

### PR DESCRIPTION
**Pull request by ulysses4ever**
As announced on the meeting, this PR adds an automated GitHub prerelease holding the `cabal-install` binaries in this repository.  If you wanna taste what it looks like, you can head out to my fork: the [Releases](https://github.com/ulysses4ever/cabal/releases) tab has the [`cabal-head`](https://github.com/ulysses4ever/cabal/releases/tag/cabal-head) prerelease with three binaries for the three main OSes in the `Assets` section -- that's what you get with this PR.

I add a note about it in the README but perhaps some more advertisement could be done (not sure how). Prereleases appear only on the Releases tab in GitHub UI, while proper releases are also featured on the index page of the repo. So, it's not very visible...

I'm not sure if `ghcup` can be used to easily get these. Suggestions are welcome. I don't think `ghcup` can add much on top of a `wget` for the direct link (e.g. `wget https://github.com/ulysses4ever/cabal/releases/download/cabal-head/cabal-head-Linux-x86_64.tar`; well, you need to untar and put it in the right spot, I guess...). Again, if someone is willing to explore it, I can add instructions to the README.

The artifacts that I put in the prerelease are built in the CI as usual, so they're debug builds with assertions enabled. It'd be more work to have actual release-mode binaries, and I didn't feel like it's worth it. Hopefully, after #8821, it's not as hard to use debug binaries. But if we get too much complains about failing assertions, we can consider going the full-blown release mode instead of debug mode.

Should close #8541, I think.

---

Please include the following checklist in your PR:

* [x] N/A ~~Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).~~
* [x] N/A ~~Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).~~
* [x] The documentation has been updated, if necessary.
* [x] N/A ~~Include [manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) if your PR relates to cabal-install.~~

### CODE REVIEW

**Comment by ulysses4ever** [comment]
@Mikolaj thanks! Yes, creating a job for ghcup metadata channel would be even better, but I don’t think it necessarily makes this PR useless. I imagine that there would be people who find GitHub prerelease UX more convenient (Also, ghcup is not always applicable, e.g. it doesn’t work on NixOS). 

Also, it’s not exactly trivial to create a ghcup channel from scratch. It’s not hard either, but my hunch is that we may have to wait for a while until this happens…

File: .github/workflows/validate.yml [hunk 1]
```diff
@@ -136,8 +136,9 @@ jobs:
              # see https://github.com/actions/virtual-environments/issues/2619#issuecomment-788397841
              sudo /usr/sbin/purge
           fi
-          tar -cvf cabal-head.tar -C $(dirname "$CABAL_EXEC") $(basename "$CABAL_EXEC")
-          echo "CABAL_EXEC_TAR=cabal-head.tar" >> $GITHUB_ENV
+          export CABAL_EXEC_TAR="cabal-head-${{ runner.os }}-x86_64.tar"
```
Comment (ffaf1) [inline .github/workflows/validate.yml:139]: Should it be `.tar` or `.tar.gz` (or `.tar.whatever-else`)?

Comment (ulysses4ever) [reply .github/workflows/validate.yml:139]: That's a good point. I just checked that `.gz` shrinks the size from 53M to 10M. There's a worry though that Windows users may not have a great support for `.gz`. (I'm not sure about `.tar` either). People used to create `zip`s just based on this in my experience. And I don't know anything about Mac. So, I just left whatever was there before (which was a tar)... Maybe if we have tar, gz shouldn't be a big additional hurdle, and I should do it, thanks!

Comment (ulysses4ever) [reply .github/workflows/validate.yml:139]: It looks like `tar`ring was added by @jneira who is also our main Windows expert. Javier, do you have any thoughts on `gz`ipping the binary on top of `tar`?

Comment (ffaf1) [reply .github/workflows/validate.yml:139]: I thought the file was already gzipped, just with the wrong extension (below we use `tar -xf …`).

Comment (ulysses4ever) [reply .github/workflows/validate.yml:139]: `tar -xf` means "eXtract a tar File". In contrast, `tar -xzf` would mean "eXtract a gZipped tar file`.

Fun fact: last week I learned that you can omit `-` in `tar -xf` and alike. How cool is that?!...

**Review by ffaf1 — APPROVED** [review]
let’s try this!

**Comment by ulysses4ever** [comment]
Let's wait a little bit and see we hear from Javier about `gzip`ping. Thanks a lot for the review!

File: .github/workflows/validate.yml (hunk 1 above)
Comment (jneira) [reply .github/workflows/validate.yml:139]: iirc there was not specific reason to not use gzip and windows support is similar between gz.tar and tar (nowadays people uses 7z which supports both)

Comment (ulysses4ever) [reply .github/workflows/validate.yml:139]: thanks! will update the pr

**Pull request merged**